<html>
<head>
	<title>ShortCut Demo Server</title>
	<meta charset="utf-8"/>
	<script type="text/javascript" src="{{ url_for('static',filename='jquery-3.3.1.js') }}"></script>
	<script type="text/javascript" src="{{ url_for('static',filename='socket.io.dev.js') }}"></script>
	<script type="text/javascript" src="{{ url_for('static',filename='index.js') }}"></script>
	<link rel= "stylesheet" type= "text/css" href= "{{ url_for('static',filename='index.css') }}">
</head>

<body>

<div id="content" style="height: 100%;">
	<div id="left" style="text-align: center;">
		<div style="width: 95%; display: inline-block;">
			<div> <!-- TODO why another div besides left? -->
				<div id="logodiv" style="display: inline-flex; align-items: center;">
					<div style="text-align: left;">
						<div id="title">"The Zilputer"<br/>ShortCut Demo Server</div>
						<pre><div id="serverinfo">users:<br/>cpu:</br>memory:</div></pre>
					</div>
					&nbsp; &nbsp;
					<img id="logo" src="/static/server.png" style="width: 125px;">
				</div>
				<br/>
				<br/>
				<div style="border-style: solid; border-width: 1px;">
					<!-- TODO how to make height auto-expand? -->
					<div id="replstdout" style="text-align: left; width:100%; height: 60%; overflow-y: scroll;"></div>
					<!-- TODO this could be one run/kill button at the end of the entry line that toggles when a script is running -->
					<div>
						<input id="replstdin" type="text" style="width: 90%;" placeholder=">>" autofocus style="float: left;">
						<button id="runstop" style="display: inline; text-align: center;">Run</button>
					</div>
				</div>
				<br/>
				<div style="display: inline-block; width: 100%;">
					<form id="script">

						<!-- TODO group these by top row being autofills and bottom being server actions -->

						<!-- load/upload -->
						<div style="float: left; width: 45%; text-align: left;">
							<button id="loadbutton">Load script</button>
							<select id="loadmenu">
								<option value="userscript1.cut">userscript1.cut</option>
								<option value="johan.cut">johan.cut</option>
								<option value="another.cut">another.cut</option>
							</select>
							<br/>
							<br/>
							<button id="uploadbutton">Upload file</button>
							<input id="upload" type="file" accept="*">
						</div>

						<div style="float: right; width: 45%; text-align: left;">
							<!-- save/download -->
							<button id="savebutton">Save script</button>
							<input id="filename" type="text" name="scriptname" placeholder="your-script-name.cut">
							<br/>
							<br/>
							<input id="dlscript" type="submit" value="Download script">
							<input id="dlresult" type="submit" value="Download result">
						</div>

					</form>
				</div>
				<br/>
				<br/>

				<div>
						<textarea id="commentfield" placeholder="Leave Jeff a question/comment/bug report! Paste anything relevant here too, and include your name if you want a response." style="width:100%; height: 5%; expand: vertical; resize: vertical; float: left;"></textarea>
						<button id="commentbutton" style="float: right;">Submit comment</button>
				</div>
			</div>
		</div>
	</div>
	<div id="right">
		<div style="width: 95%;">
{% filter markdown %}

Phylogenomic cuts are lists of candidate genes whose distribution suggests they
might be imprortant for a trait of interest.

Cut script is a simple(ish) language for making phylogenomic cuts. It aims to
be the easiest way to conduct searches too complex for the NCBI website, but
not so complex that you want to get bogged down with all the details of a
"real" language like Python or R just to do some BLAST searches. It also aims
to describe searches succinctly and reproducibly, so that some day it might
make it into your supplemental material. Everyone wins when your work is easy
to build on!

The `shortcut` program is a command-line language interpreter, kind of like
`python` or `R` but for cut scripts. On your left is a hacky "terminal" I made
so it can be run from a web browser. You can do most of the same things in it
that you could if you downloaded ShortCut proper, except:

* Your session will be deleted when you leave this page
* The server might kill your script without warning if it takes too much CPU time
* Some terminal niceties like tab completion and clearing the screen are missing

You can upload your own genomes and gene lists, save and restore scripts,
and download results using the buttons at the bottom left. *Note: buttons don't work yet*

## Quick Start

OK, so what can you do with it? These examples are explained in more detail in
the tutorial below, but let's try running them first. Press one of the `Load`
buttons to fill in the command to load that example. Once it's loaded type
`:show` to show the variable definitions, then `result` to run everything.

{% endfilter %}
{% with path='examples/quick01.cut' %} {% include "example.html" %} {% endwith %}

{% with path='examples/quick02.cut' %} {% include "example.html" %} {% endwith %}

{% with path='examples/diatom.cut' %} {% include "example.html" %} {% endwith %}
{% filter markdown %}

_Note: these mostly use BLAST, but multiple sequence alignments +
tree building + clustering are in the works too._

## Tutorial

The best way to use this tutorial is probably to
load each example and play around with it a bit as you go. That's the fastest
way to learn, and also the fastest way for me to get bug reports. Just type
them in the box on the lower left. Something short like "I did this and
expected this, but this happened instead" is fine. If the tutorial is
confusing, that counts as a bug too!

### `result` and other variables

Let's start at the beginning.
This is probably the simplest script you could write:

{% with path='examples/variables01.cut' %} {% include "example.html" %} {% endwith %}

If you downloaded ShortCut and ran `shortcut --script examples/variables01.cut`, it would print
`"hello world!"`.

You don't have to run a whole script at once though.
You'll spend most of your time editing it in the interpreter,
defining and evaluating individual variables.
Here is a script with several of them.

{% with path='examples/variables02.cut' %} {% include "example.html" %} {% endwith %}

ShortCut keeps track of dependencies between variables, like this:

![]({{ url_for('static',filename='vars.svg') }})

You can evaluate one of them by typing its name.
Anything it depends on will also get evaluated.
You can find what a given variable depends on with the `:depends` command,
and what depends on it with `:rdepends` ("reverse dependencies").

`result` always holds the latest result.
If you type a plain expression like `4 * 4` without assigning it to a variable,
that becomes the new `result`.
You can also assign it yourself when the script is done to say what the final result should be.

So, if you type `result` or `var4` here the graph will stay the same and
ShortCut will print `-1.485`. Then if you type `var2`, the graph will change to:

![]({{ url_for('static',filename='var2.svg') }})

... and it will print `"nothing depends on var2 so far"`.

### Math

Math is pretty simple. It works like you would expect, except that everything
is left-to-right rather than following order of operations. _Note: that could
be added easily if people prefer it._

A few examples:

{% with path='examples/math01.cut' %} {% include "example.html" %} {% endwith %}

You can enter numbers in decimal or scientific notation. There's only one type
of number instead of the several different ones like doubles and floats you
would find in a typical language. Parentheses also work the regular "mathy" way
to group things together; you'll see a little further down that function
application doesn't use them.

Notice that ShortCut might remove your parentheses automatically,
but only where it doesn't change the meaning.
<!-- TODO bug: it actually does seem to matter here. Put PEMDAS back? -->

### Sets

Besides regular math, you might want to do set operations. They also apply left
to right and can be grouped with parentheses. They're very useful for comparing
sets of genes and genomes. With these 3 you can describe the equivalent of any
Venn diagram:

_Note: actual Venn diagrams coming soon._

### Types

<!-- TODO bug: brackets misplaced in :type of & -->

### NCBI BLAST+

ShortCut includes the most common command line BLAST tools:

* `blastn`
* `blastp`
* `blastx`
* `tblastn`
* `tblastx`
* `megablast`
* `psiblast`

_Note: Just ask if you want one of the more exotic ones added, like `rps-blast` or `delta-blast`._

_Note: `:help` for individual functions coming soon._

### PSI-BLAST

When making a cut you'll often want to do a sensitive search for distant
homologs, and PSI-BLAST (position-specific iterated blast) is usually the best
one for that. You do have to be a little careful though, because it's possible
for the search to be "poisoned": if the first BLASTP iteration picks up
unrelated proteins, and they have more hits in the database than your actual
protein of interest does, it will train itself to pick up more and more of the
wrong ones.

So if you only have one or a few genes to search for the best strategy is to
look through the results of each iteration on the NCBI site to make sure they
seem reasonable. Obviously that's not very high-throughput though. ShortCut
can't assess whether the results make sense, but you can use it to tune the
search settings to pick up known positive control genes while excluding as many
others as possible. That's the topic of the next section.

If you have a large number of similar genes, like a bunch of membrane
transporters, you might also try a hybrid strategy:

1. pick a few random ones with the `sample` function
2. confirm those work on the NCBI site
3. do them all in ShortCut

### PRS pattern

## TODO

Tutorial:

* Transcribe from poster
* Tmpfiles
* Types (genes, genomes, lists)
* Sets (lists, subtraction)
* BLAST+ (types, functions)
* Psiblast (pssms, functions)
* PRS Pattern
* Permute (drop each, sample)
* Repeat (repeat and randomness, repeat_each)
* Summarize (length_each, score_*, plot_*)

{% endfilter %}
		</div>
	</div>
</div>
	</body>
</html>
